src/main/java/org/folio/service/inventory/InventoryInstanceManager.java

```diff
@@ -347,17 +343,46 @@ private Future<String> createInstance(JsonObject instanceRecJson, RequestContext
   public Future<CompositePoLine> openOrderHandleInstance(CompositePoLine compPOL, boolean isInstanceMatchingDisabled, RequestContext requestContext) {
     return consortiumConfigurationService.getConsortiumConfiguration(requestContext)
       .compose(consortiumConfiguration -> {
-        if (compPOL.getInstanceId() != null) {
-          return consortiumConfiguration.isEmpty() ? Future.succeededFuture(compPOL.getInstanceId())
-            : shareInstanceAmongTenantsIfNeeded(compPOL.getInstanceId(), consortiumConfiguration.get(), compPOL.getLocations(), requestContext);
-        } else {
-          return getInstanceRecord(compPOL, isInstanceMatchingDisabled, requestContext)
-            .compose(instanceId -> consortiumConfiguration.isEmpty() ? Future.succeededFuture(instanceId)
-              : shareInstanceAmongTenantsIfNeeded(instanceId, consortiumConfiguration.get(), compPOL.getLocations(), requestContext));
-        }
+        Future<String> instanceIdFuture = compPOL.getInstanceId() != null ?
+          Future.succeededFuture(compPOL.getInstanceId()) :
+          getInstanceRecord(compPOL, isInstanceMatchingDisabled, requestContext);
+        consortiumConfiguration.ifPresent(
+          configuration -> instanceIdFuture
+            .compose(instanceId -> shareInstanceAmongTenantsIfNeeded(instanceId, configuration, compPOL.getLocations(), requestContext))
+        );
+        return instanceIdFuture;
       }).map(compPOL::withInstanceId);
   }
 
+  private Future<String> shareInstanceAmongTenantsIfNeeded(String instanceId, ConsortiumConfiguration consortiumConfiguration,
+                                                           List<Location> locations, RequestContext requestContext) {
+    return findTenantsWithUnsharedInstance(instanceId, locations, requestContext)
+      .map(tenantIds -> tenantIds.stream().map(targetTenantId -> sharingInstanceService.createShadowInstance(instanceId,
+        targetTenantId, consortiumConfiguration, requestContext)).toList())
+      .compose(HelperUtils::collectResultsOnSuccess)
+      .map(sharingInstances -> instanceId);
+  }
+
+  /*
+  Returns list of tenant ids for which instance wasn't found
+   */
+  private Future<List<String>> findTenantsWithUnsharedInstance(String instanceId, List<Location> locations, RequestContext requestContext) {
```

Comment:
name and comment is confusing. Unshared mean it is still exists. Probably it is okay to remove comments

Reply:
This method is old, I added a comment but I'll remove if name is self-explaining

Reply:
better to do, thanks

Reply:
done